`windows-gnu`: Adhere to MinGW convention for build outputs

Doekin opened the pull request:
This PR potentially addresses issues #20039, #21103, and #16811.

### Background
As outlined in #20039, when attempting to link a library named `foo`, the default search paths are `foo.dll`, `foo.lib`, and `libfoo.a`. However, the conventional DLL name format in MinGW is `libfoo.dll` and `libfoo.dll.a`. This mismatch can cause issues when working with build tools like CMake. This PR aims to resolve the inconsistency and improve compatibility.

### Changes
- Updated `fn fileExt(of: ObjectFormat, arch: Cpu.Arch)` in `std.Target.ObjectFormat` to `fn fileExt(of: ObjectFormat, abi: Abi, arch: Cpu.Arch)`.
- Changed default output names for MinGW targets:
  - Object files: `.obj` → `.o`
  - Static libraries: `foo.lib` → `libfoo.a`
  - Shared libraries: `foo.dll` → `libfoo.dll`, import libraries: `foo.lib` → `libfoo.dll.a`
  - Debug symbols: `.pdb` → DWARF format.
- Adjusted library search paths to reflect the new output names.
- Default behavior updated to not emit import libraries for executables or shared libraries when targeting the GNU ABI.
- Added support for the `-mwindows` linker flag (complements #11396).

### Example
GCC/Clang generate a 66KiB file when creating a simple shared library. Previously, `zig 0.13` would generate three files: `a.exe` (144.5 KiB), `a.pdb` (872.0 KiB), and `foo.lib` (1.2 KiB). After these changes, `zig cc` generates a single file: `a.exe` (874.0 KiB).

Additionally, when attempting to link to a non-existent library, the error message now reflects the updated search order:

#### Before
```
$ zig cc -target x86_64-windows-gnu main.c -lfoo -L.
error: unable to find dynamic system library 'foo' using strategy 'paths_first'. searched paths:
  .\foo.dll
  .\foo.lib
  .\libfoo.a
```

#### After
```
$ zig cc -target x86_64-windows-gnu main.c -lfoo -L.
error: unable to find dynamic system library 'foo' using strategy 'paths_first'. searched paths:
  .\libfoo.dll
  .\libfoo.dll.a
  .\foo.dll
  .\libfoo.a
  .\foo.lib
```

Doekin committed:
windows-gnu: Adhere to MinGW convention for debug format and object file extension.

Doekin committed:
windows-gnu: Add support for `-mwindows` option.

Doekin committed:
windows-gnu: Enhance MinGW library resolution to include additional file formats

alexrp commented:
> Disabled default import library emission in `zig cc` mode (still enabled by default in `zig build`).

This is the only part of this PR for which the rationale is not immediately obvious to me. Can you elaborate on this point?

Doekin commented:
Thank you for your feedback. 

When using `zig cc` as a replacement for LLVM-MinGW, I noticed that `main.lib` was generated alongside `HelloQt.exe`. Since GCC and Clang do not emit import libraries by default, I felt it would be more consistent to align `zig cc` with this behavior.

alexrp commented:
Ok, that makes sense. I would suggest including that rationale in the commit message.

alexrp commented on lib/std/Target.zig:
```diff
@@ -1033,10 +1035,10 @@ pub const ObjectFormat = enum {
     // - dxcontainer
 
-    pub fn fileExt(of: ObjectFormat, arch: Cpu.Arch) [:0]const u8 {
+    pub fn fileExt(of: ObjectFormat, abi: Abi, arch: Cpu.Arch) [:0]const u8 {
```
Given how much `Target` data this function now needs, I think it's time to consider adding an `objFileExt()` helper function on `Target`, similar to what's done for the other functions of this kind.

alexrp commented on src/main.zig:
```diff
@@ -473,6 +473,7 @@ const usage_build_generic =
     \\  -fno-structured-cfg       (SPIR-V) force SPIR-V kernels to not use structured control flow
     \\  -mexec-model=[value]      (WASI) Execution model
     \\  -municode                 (Windows) Use wmain/wWinMain as entry point
+    \\  -mwindows                 (Windows) Specifies a GUI application is to be generated
```
Have you verified that this flag has no effects on e.g. the preprocessor? If it does, then it must be passed along to Clang in `Compilation.addCCArgs()` just as we do for `-municode`.

Doekin replied:
According to [GCC’s documentation](https://gcc.gnu.org/onlinedocs/gcc/Cygwin-and-MinGW-Options.html#index-mwindows), the `-mwindows` flag tells the linker to set the PE header subsystem to `WINDOWS`. Clang’s source also shows that `-mwindows` adds `--subsystem windows` to the linker. You can see the relevant code here: https://github.com/llvm/llvm-project/blob/8d2b070f07969f8358c93b9cf56834806fdc3efa/clang/lib/Driver/ToolChains/MinGW.cpp#L148-L153.

When we build a GUI program in CMake with `add_executable(my_exe WIN32 main.c)`, `-mwindows` is applied only at link time for executables. So, it appears OK to not pass `-mwindows` to Clang.

Doekin committed:
Target: Add `objFileExt()` convenience wrapper

Added the `objFileExt()` function to Target, streamlining access to target-specific data in a manner consistent with other convenience functions.

alexrp commented on src/Compilation/Config.zig:
```diff
@@ -444,9 +444,10 @@ pub fn resolve(options: Options) ResolveError!Config {
         break :b switch (target.ofmt) {
             .elf, .goff, .macho, .wasm, .xcoff => .{ .dwarf = .@"32" },
-            .coff => .code_view,
+            .coff => if (target.abi.isGnu()) .{ .dwarf = .@"32" } else .code_view,
```
This change -- DWARF by default (i.e. embedded into the binary; no PDB) for MinGW -- is probably the only controversial change in this PR. I don't personally have a problem with it since you can still opt into getting a PDB, but we should probably think carefully about this change nonetheless.

cc @squeek502 @andrewrk for thoughts, maybe others?

squeek502 replied:
I don't feel like I know enough about why MinGW made this choice to give any useful feedback here.

I have a more general question, though (and this is also coming from ignorance, this is not necessarily intended as a criticism of these changes): is the stuff in this PR related to 'targeting the MinGW ABI' or is it more related to 'emulating the MinGW toolchain'? If it's the latter, is that something that Zig-the-compiler is interested in?

kcbanner replied:
Native debuggers on Windows all expect PDBs, I think this change would hurt debuggability there - it would mean users would have to always specify a non-default option to use the default platform debuggers.

alexrp replied:
An interesting data point for this discussion: https://github.com/msys2/MINGW-packages/issues/5646

squeek502 replied:
> An interesting data point for this discussion: [msys2/MINGW-packages#5646](https://github.com/msys2/MINGW-packages/issues/5646)

This is also mentioned in https://github.com/ziglang/zig/issues/20039:

> Zig uses PDB or CodeView format for debug symbols, while Rust and GCC use DWARF
> 
> * When using Zig library in Rust or C on Windows, the mix of debug symbol formats makes either language un-debuggable
> * GDB has no support of PDB files (which use CodeView debug format) and LLDB's support is problematic ([[1]](https://github.com/vadimcn/codelldb/wiki/Windows), [[2]](https://werat.dev/blog/debugging-wine-with-lldb-and-vscode/#:~:text=There%E2%80%99s%20one%20caveat,only%20on%20Windows.)), which makes open-sourced VS Code debuggers like [CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb) unusable for zig debugging purposes

I'm not sure there's a default that's "good" here. `gnu` being the default ABI for Windows in Zig makes things a bit tricky.

Doekin replied:
> Is the stuff in this PR related to 'targeting the MinGW ABI' or is it more related to 'emulating the MinGW toolchain'?

It's more about the latter. The goal of this PR is to improve compatibility.

alexrp commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -651,6 +657,7 @@ pub fn producesPdbFile(compile: *Compile) bool {
         .windows, .uefi => {},
         else => return false,
     }
+    if (target.abi.isGnu()) return false;
```
There still has to be a way to opt into getting a CodeView PDB for MinGW.

This'll mean adding `-gcodeview` support, which appears to be missing at the moment. But if you take a look at the `-gdwarf32`/`-gdwarf64` handling in the compiler driver, it should be obvious how to add it.

As far as `std.Build` goes, you'll need to replace the `Module.dwarf_format` field with a `debug_format` field, probably based on the compiler's `Compilation.Config.DebugFormat` union, which will need to be moved to the standard library. I suggest putting it in `std.builtin` for now. (I have been thinking about making it part of `std.Target` like `ObjectFormat`, but that'll take deeper surgery, so let's not do that here.) Once this is done, this function can check the root module's `debug_format` and return a more accurate result.

Doekin replied:
At [@3867570](https://github.com/ziglang/zig/pull/22415/commits/38675707183a345b0151dd23cbcf13f2b411759a), `zig.exe cc -target x86_64-windows-gnu hello.c` produces an 891.0 KiB binary, and `zig.exe cc -target x86_64-windows-gnu hello.c -gcodeview` produces an 847.0 KiB binary and a 224.0 KiB PDB. 

It confuses me that DWARF debug info is still included with `-gcodeview`. Using Zig 0.13, `zig cc -target x86_64-windows-gnu hello.c` outputs a 148.5 KiB binary and a 1004.0 KiB PDB.

alexrp replied:
I suspect the issue is that we're not passing the debug format given on the command line through to the sub-compilation for `mingw32.lib`. The same is probably true for compiler-rt as well. If you grep for `root_strip` in `src/Compilation.zig`, it should be easy enough to fix.

Doekin replied:
Thanks, it's working now.

Doekin committed:
Debug: Move `Compilation.Config.DebugFormat` to `std.builtin` for broader usage

Doekin committed:
Debug: Replace `Module.dwarf_format` with `std.builtin.debug_format` for enhanced flexibility

alexrp commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -651,6 +657,8 @@ pub fn producesPdbFile(compile: *Compile) bool {
         .windows, .uefi => {},
         else => return false,
     }
+    if (compile.root_module.debug_format == null or
+        compile.root_module.debug_format.? != .code_view) return false;
```
Note that null means "use compiler default", so this is not quite right.

Doekin replied:
Got it. Now, if `debug_format` is null and the ABI is GNU, `producesPdbFile` returns false.

Doekin committed:
windows-gnu: Add support for `-gcodeview` flag

Doekin committed:
build: Disable default import library emission for `.exe` and MinGW `.dll`

Aligns with other compilers' behavior.

Doekin committed:
link: Simplify library path resolution logic

inschrift-spruch-raum commented:
How is the current progress?

andrewrk commented:
This pull request is not ready for review because:
* It has conflicts that must be resolved via rebasing against latest origin/master.
* It is not passing the CI tests.

Closing since there has been no activity for 30+ days.
